README.md
# upn_qr

[![pub package](https://img.shields.io/pub/v/upn_qr)](https://pub.dev/packages/upn_qr)
[![Analyze & Test](https://github.com/greymag/dart_upn_qr/actions/workflows/dart.yml/badge.svg?branch=main)](https://github.com/greymag/dart_upn_qr/actions/workflows/dart.yml)
[![innim lint](https://img.shields.io/badge/style-innim_lint-40c4ff.svg)](https://pub.dev/packages/innim_lint)

Library for parse UPN QR code content.

UNP (Univerzalni plačilni nalog) - Universal Payment Order for Slovenia.

See https://www.zbs-giz.si/placilni-promet/

## Usage

Use methods from `UpnQRParser`.

```dart
const data = UpnQRParser.parseString(qrCodeString);
print('Recipient IBAN: ${data.recipientIban}, amount: ${data.amount}')
```

CHANGELOG.md
## 0.0.5

- Methods to get dates as `DateTime`.
- Method to get amount as `int`.

## 0.0.4

- Two `UpnData` object with the same fields values considered as equals.

## 0.0.3

- Added `toString()` for `UpnData`.
- Some documentation for classes and enums
- README: added badges.
- Added example.

## 0.0.2

Verification and checks.

## 0.0.1

Very first basic version of UPN QR code content.


test/upn_data_test.dart
import 'package:test/test.dart';
import 'package:tuple/tuple.dart';
import 'package:upn_qr/upn_qr.dart';

void main() {
  group('UpnData', () {
    group('equality:', () {
      test('should be equal to itself', () {
        final data = createData();

        expect(data == data, true);
      });

      test('should be equal to another object with same field values', () {
        final a = createData();
        final b = createData();

        expect(a == b, true);
      });

      test('should be not equal to another object with different payerIban',
          () {
        final a = createData(payerIban: 'SI56011008881234567');
        final b = createData(payerIban: 'BE56011008881234567');

        expect(a == b, false);
      });

      test('should be not equal to another object with different deposit', () {
        final a = createData(deposit: true);
        final b = createData(deposit: false);

        expect(a == b, false);
      });

      test('should be not equal to another object with different withdrawal',
          () {
        final a = createData(withdrawal: true);
        final b = createData(withdrawal: false);

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different payerReference',
          () {
        final a = createData(payerReference: 'RF035126080');
        final b = createData(payerReference: 'RF123456789');

        expect(a == b, false);
      });

      test('should be not equal to another object with different payerName',
          () {
        final a = createData(payerName: '1');
        final b = createData(payerName: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different payerAddress',
          () {
        final a = createData(payerAddress: '1');
        final b = createData(payerAddress: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different payerCity',
          () {
        final a = createData(payerCity: '1');
        final b = createData(payerCity: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different amount', () {
        final a = createData(amount: '00000003400');
        final b = createData(amount: '00000001200');

        expect(a == b, false);
      });

      test('should be not equal to another object with different paymentDate',
          () {
        final a = createData(paymentDate: null);
        final b = createData(paymentDate: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different urgent', () {
        final a = createData(urgent: true);
        final b = createData(urgent: false);

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different paymentPurposeCode',
          () {
        final a = createData(paymentPurposeCode: 'VAT');
        final b = createData(paymentPurposeCode: 'SAL');

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different paymentPurpose',
          () {
        final a = createData(paymentPurpose: '1');
        final b = createData(paymentPurpose: '2');

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different paymentDeadline',
          () {
        final a = createData(paymentDeadline: '31.08.2023');
        final b = createData(paymentDeadline: '30.08.2023');

        expect(a == b, false);
      });

      test('should be not equal to another object with different recipientIban',
          () {
        final a = createData(recipientIban: 'SI56011008881000030');
        final b = createData(recipientIban: 'SI12345008881000030');

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different recipientReference',
          () {
        final a = createData(recipientReference: '1');
        final b = createData(recipientReference: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different recipientName',
          () {
        final a = createData(recipientName: '1');
        final b = createData(recipientName: '2');

        expect(a == b, false);
      });

      test(
          'should be not equal to another object with different recipientAddress',
          () {
        final a = createData(recipientAddress: '1');
        final b = createData(recipientAddress: '2');

        expect(a == b, false);
      });

      test('should be not equal to another object with different recipientCity',
          () {
        final a = createData(recipientCity: '1');
        final b = createData(recipientCity: '2');

        expect(a == b, false);
      });
    });

    group('getPaymentDate()', () {
      test('should return correct DateTime if presented', () {
        final data = createData(paymentDate: '11.07.2016');

        final res = data.getPaymentDate();

        expect(res, DateTime(2016, 7, 11));
      });

      test('should return null if not presented', () {
        final data = createData(paymentDate: null);

        final res = data.getPaymentDate();

        expect(res, null);
      });
    });

    group('getPaymentDeadline()', () {
      test('should return correct DateTime if presented', () {
        final data = createData(paymentDeadline: '06.11.2023');

        final res = data.getPaymentDeadline();

        expect(res, DateTime(2023, 11, 6));
      });

      test('should return null if not presented', () {
        final data = createData(paymentDeadline: null);

        final res = data.getPaymentDeadline();

        expect(res, null);
      });
    });

    group('getPaymentDeadline()', () {
      const [
        Tuple2('00000126874', 126874),
        Tuple2('00000018500', 18500),
        Tuple2('00000003396', 3396),
      ].forEach((e) {
        final amount = e.item1;
        final expected = e.item2;
        test('should return correct value for <$amount>', () {
          final data = createData(amount: amount);

          final res = data.getAmount();

          expect(res, expected);
        });
      });

      test('should return 0 if value is 0', () {
        final data = createData(amount: '00000000000');

        final res = data.getAmount();

        expect(res, 0);
      });

      test('should return 0 if value is empty', () {
        final data = createData(amount: '');

        final res = data.getAmount();

        expect(res, 0);
      });
    });
  });
}

UpnData createData({
  String? payerIban = 'SI56011008881234567',
  bool deposit = false,
  bool withdrawal = false,
  String? payerReference = 'RF035126080',
  String? payerName = 'ORANGE DOG',
  String? payerAddress = 'BARJANSKA CESTA 70',
  String? payerCity = '1000 Ljubljana',
  String amount = '00000003400',
  String? paymentDate,
  bool urgent = false,
  String? paymentPurposeCode = 'VAT',
  String? paymentPurpose = 'Payment by invoice',
  String? paymentDeadline = '31.08.2023',
  String? recipientIban = 'SI56011008881000030',
  String? recipientReference = 'SI1935980621-62006',
  String? recipientName = 'Receiver',
  String? recipientAddress = 'Main street 20',
  String? recipientCity = 'Ljubljana',
}) =>
    UpnData(
      payerIban: payerIban,
      deposit: deposit,
      withdrawal: withdrawal,
      payerReference: payerReference,
      payerName: payerName,
      payerAddress: payerAddress,
      payerCity: payerCity,
      amount: amount,
      paymentDate: paymentDate,
      urgent: urgent,
      paymentPurposeCode: paymentPurposeCode,
      paymentPurpose: paymentPurpose,
      paymentDeadline: paymentDeadline,
      recipientIban: recipientIban,
      recipientReference: recipientReference,
      recipientName: recipientName,
      recipientAddress: recipientAddress,
      recipientCity: recipientCity,
    );


test/upn_qr_parser_test.dart
import 'package:test/test.dart';
import 'package:upn_qr/upn_qr.dart';

part 'upn_qr_parser_test.data.dart';
part 'upn_qr_parser_test.expect.dart';

void main() {
  group('UpnQRParser', () {
    group('parseString()', () {
      test('should parse string content', () {
        final res = UpnQRParser.parseString(upnQR_1);
        expectBasicData(res);
      });
      test('should parse string content with space at end of address', () {
        final res = UpnQRParser.parseString(upnQR_WithSpaceAtEndOfAddress);
        expectBasicData(res, payerAddress: 'BARJANSKA CESTA 70 ');
      });
      test('should parse string content sample', () {
        final res = UpnQRParser.parseString(upnQR_Sample);

        expectSample(res);
      });

      test(
          'should parse string example A: '
          'Order for the payment of obligations (cash)', () {
        final res = UpnQRParser.parseString(upnQR_ExampleA);
        expectExampleA(res);
      });

      test(
          'should parse string example B: '
          'Order for the payment of obligations (non-cash)', () {
        final res = UpnQRParser.parseString(upnQR_ExampleB);
        expectExampleB(res);
      });

      test('should parse zero amount', () {
        final res = UpnQRParser.parseString(upnQR_ZeroAmount);
        expectBasicData(res, amount: '00000000000');
      });

      test('should parse content with reserve string', () {
        final res = UpnQRParser.parseString(upnQR_21_lines);
        expectBasicData(res);
      });

      test('should throw invalidFormat error if no leading', () {
        expect(
          () => UpnQRParser.parseString(upnQR_NoLeading),
          _throwsA(UnpQRParseErrorType.invalidFormat),
        );
      });

      test('should throw tooLong error if length in more that 411', () {
        expect(
          () => UpnQRParser.parseString(upnQR_412_chars),
          _throwsA(UnpQRParseErrorType.tooLong),
        );
      });

      test(
          'should throw tooFewLines error if '
          'content has more than 19 lines', () {
        expect(
          () => UpnQRParser.parseString(upnQR_19_lines),
          _throwsA(UnpQRParseErrorType.tooFewLines),
        );
      });

      test(
          'should throw tooManyLines error if '
          'content has more than 21 lines', () {
        expect(
          () => UpnQRParser.parseString(upnQR_22_lines),
          _throwsA(UnpQRParseErrorType.tooManyLines),
        );
      });

      test(
          'should throw invalidChecksumFormat error if check sum is not a number',
          () {
        expect(
          () => UpnQRParser.parseString(upnQR_InvalidChecksumFormat),
          _throwsA(UnpQRParseErrorType.invalidChecksumFormat),
        );
      });

      test(
          'should throw checksumNotMatched error if '
          'check sum is not matched content', () {
        expect(
          () => UpnQRParser.parseString(upnQR_InvalidChecksumValue),
          _throwsA(UnpQRParseErrorType.checksumNotMatched),
        );
      });

      test('should throw missedAmount error if no amount', () {
        expect(
          () => UpnQRParser.parseString(upnQR_NoAmount),
          _throwsA(UnpQRParseErrorType.missedAmount),
        );
      });

      test('should throw missedAmount error if no checksum', () {
        expect(
          () => UpnQRParser.parseString(upnQR_InvalidChecksumEmpty),
          _throwsA(UnpQRParseErrorType.missedChecksum),
        );
      });
    });
  });
}

Matcher _throwsA(UnpQRParseErrorType type) =>
    throwsA(isA<UnpQRParseError>().having((e) => e.type, 'type', type));


test/upn_qr_parser_test.data.dart
// ignore_for_file: constant_identifier_names

part of 'upn_qr_parser_test.dart';

const upnQR_1 = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262
''';

const upnQR_WithSpaceAtEndOfAddress = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70 
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
263
''';

const upnQR_Sample = '''
UPNQR
SI56020170014356205


SI00225268-32526-222
Any company Ltd.
Lepa cesta 10
2000 Maribor
00000126874
11.07.2016
X
CPYR
Invoice payment 18/2016.

SI56020360253863406
SI081236-17-345679
Bank Association of Slovenia - GIZ
Šubičeva 2
1000 Ljubljana
249''';

// Example of a completed UPN QR form -
// Order for the payment of obligations (cash)
const upnQR_ExampleA = '''
UPNQR




Janez Novak
Dunajska 1
1000 Ljubljana
00000008105


COST
Plačilo obveznosti 10/2016

SI56051008010486080
SI0598765432100
Novo podjetje d.o.o.
Lepa cesta 15
3698 Loški Potok
183
''';

// Example of a completed UPN QR form -
// Order for the payment of obligations (non-cash)
const upnQR_ExampleB = '''
UPNQR
SI56020170014356205


SI003528-990
Združenje bank Slovenije
Šubičeva 2
1000 Ljubljana
00000128067


ADVA
Plačilo avansa-ponudba 2016/12

SI56051008010486080
SI00123456-67890-12345
Novo podjetje d.o.o.
Lepa cesta 15
3698 Loški Potok
238''';

const upnQR_ZeroAmount = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000000000


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262''';

const upnQR_NoLeading = '''
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


Some
VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262
''';

const upnQR_NoAmount = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana



VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262''';

const upnQR_19_lines = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
262''';

const upnQR_21_lines = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262
      ''';

const upnQR_22_lines = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
262

12''';

const upnQR_412_chars = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMIORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70BARJANSKA CESTA 70BARJANSKA CESTA 70BARJANSKA
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUNRS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20Gregorčičeva ulica 20Gregorčičeva ulica 20
Ljubljana
262''';

const upnQR_InvalidChecksumValue = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
263''';

const upnQR_InvalidChecksumFormat = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
26o''';

const upnQR_InvalidChecksumEmpty = '''
UPNQR
SI56011008881234567


RF035126080
ORANGE DOG, RAČUNALNIŠKO PROGRAMI
BARJANSKA CESTA 70
1000 Ljubljana
00000003400


VATX
Plačilo DDV za: 7-7.2023
31.08.2023
SI56011008881000030
SI1935980621-62006
RS PREHODNI DAVČNI PODRAČUN
Gregorčičeva ulica 20
Ljubljana
''';


lib/src/upn_data.dart
import 'package:equatable/equatable.dart';

/// Data of UNP (Univerzalni plačilni nalog).
///
/// Universal Payment Order for Slovenia.
class UpnData extends Equatable {
  /// Payer’s IBAN
  ///
  /// SEPA IBAN without spaces.
  ///
  /// Example: SI56020170014356205
  final String? payerIban;

  /// Deposit
  ///
  /// `true` if payment is deposit.
  ///
  /// Both [deposit] and [withdrawal] can be `false`, if not defined.
  final bool deposit;

  /// Withdrawal
  ///
  /// `true` if payment is withdrawal.
  ///
  /// Both [deposit] and [withdrawal] can be `false`, if not defined.
  final bool withdrawal;

  /// Payer’s reference
  ///
  /// Model and reference together without spaces.
  ///
  /// First 4 characters - SI or RF model rules.
  /// Next 22 characters - Rules for SI or RF reference (without spaces).
  ///
  /// Example: SI00225268-32526-222
  final String? payerReference;

  /// Payer’s name
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces.
  ///
  /// Example: Any company Ltd.
  final String? payerName;

  /// Payer’s street and No.
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces.
  ///
  /// Example: Lepa cesta 10
  final String? payerAddress;

  /// Payer’s city
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces..
  ///
  /// Example: 2000 Maribor
  final String? payerCity;

  /// Amount.
  ///
  /// Just numbers with leading zeros.
  /// Amount multiplied by 100 and without prefix or decimal places.
  /// The maximum amount hat can be entered is
  /// 999,999,999,999.99 (one billion - 1).
  /// If 0, the content is "00000000000"
  ///
  /// Example: 00000126874 (for 1.268,74).
  final String amount;

  /// Payment date
  ///
  /// Format 'DD.MM.YYYY'.
  ///
  /// Example: 11.07.2016
  final String? paymentDate;

  /// Urgent,
  ///
  /// `true` if payment if urgent.
  ///
  /// Example:
  final bool urgent;

  /// Purpose code.
  ///
  /// Uppercase letters (A-Z). Purpose code rules.
  ///
  /// 4 characters length.
  ///
  /// Example: CPYR
  final String? paymentPurposeCode;

  /// Purpose of payment
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces.
  ///
  /// Example: Invoice payment 18/2016.
  final String? paymentPurpose;

  /// Payment deadline
  ///
  /// Format 'DD.MM.YYYY'.
  ///
  /// Example: 11.07.2016
  final String? paymentDeadline;

  /// Recipient’s IBAN
  ///
  /// SEPA IBAN without spaces.
  ///
  /// Example: SI56020360253863406
  final String? recipientIban;

  /// Recipient’s reference
  ///
  /// Model and reference together without spaces.
  ///
  /// First 4 characters - SI or RF model rules.
  /// Next 22 characters - Rules for SI or RF reference (without spaces).
  ///
  /// Example: SI081236-17-345679
  final String? recipientReference;

  /// Recipient’s name
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces.
  ///
  /// Example: Bank Association of Slovenia - GIZ
  final String? recipientName;

  /// Recipient’s street and No.
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces.
  ///
  /// Example: Šubičeva 2
  final String? recipientAddress;

  /// Recipient’s city
  ///
  /// Post code and the city.
  ///
  /// All characters of the ISO 8859-2 code table.
  /// No leading or following spaces..
  ///
  /// Example: 1000 Ljubljana
  final String? recipientCity;

  const UpnData({
    required this.payerIban,
    required this.deposit,
    required this.withdrawal,
    required this.payerReference,
    required this.payerName,
    required this.payerAddress,
    required this.payerCity,
    required this.amount,
    required this.paymentDate,
    required this.urgent,
    required this.paymentPurposeCode,
    required this.paymentPurpose,
    required this.paymentDeadline,
    required this.recipientIban,
    required this.recipientReference,
    required this.recipientName,
    required this.recipientAddress,
    required this.recipientCity,
  });

  int getAmount() => amount.isNotEmpty ? int.parse(amount) : 0;

  DateTime? getPaymentDate() => _convertDateTime(paymentDate);

  DateTime? getPaymentDeadline() => _convertDateTime(paymentDeadline);

  DateTime? _convertDateTime(String? value) {
    if (value == null) return null;

    final parts = value.split('.');
    if (parts.length != 3) {
      throw ArgumentError.value(value, 'value', 'Invalid value format');
    }

    final day = parts[0];
    final month = parts[1];
    final year = parts[2];

    return DateTime(int.parse(year), int.parse(month), int.parse(day));
  }

  @override
  List<Object?> get props => [
        payerIban,
        deposit,
        withdrawal,
        payerReference,
        payerName,
        payerAddress,
        payerCity,
        amount,
        paymentDate,
        urgent,
        paymentPurposeCode,
        paymentPurpose,
        paymentDeadline,
        recipientIban,
        recipientReference,
        recipientName,
        recipientAddress,
        recipientCity,
      ];

  @override
  String toString() => 'UpnData{'
      'payerIban: $payerIban, '
      'deposit: $deposit, '
      'withdrawal: $withdrawal, '
      'payerReference: $payerReference, '
      'payerName: $payerName, '
      'payerAddress: $payerAddress, '
      'payerCity: $payerCity, '
      'amount: $amount, '
      'paymentDate: $paymentDate, '
      'urgent: $urgent, '
      'paymentPurposeCode: $paymentPurposeCode, '
      'paymentPurpose: $paymentPurpose, '
      'paymentDeadline: $paymentDeadline, '
      'recipientIban: $recipientIban, '
      'recipientReference: $recipientReference, '
      'recipientName: $recipientName, '
      'recipientAddress: $recipientAddress, '
      'recipientCity: $recipientCity'
      '}';
}


lib/src/upn_qr_parse_error.dart
/// UPN QR parse error
class UnpQRParseError extends Error {
  /// Type
  final UnpQRParseErrorType type;

  UnpQRParseError(this.type);

  factory UnpQRParseError.invalidFormat() =>
      UnpQRParseError(UnpQRParseErrorType.invalidFormat);

  factory UnpQRParseError.tooLong() =>
      UnpQRParseError(UnpQRParseErrorType.tooLong);

  factory UnpQRParseError.tooFewLines() =>
      UnpQRParseError(UnpQRParseErrorType.tooFewLines);

  factory UnpQRParseError.tooManyLines() =>
      UnpQRParseError(UnpQRParseErrorType.tooManyLines);

  factory UnpQRParseError.invalidChecksumFormat() =>
      UnpQRParseError(UnpQRParseErrorType.invalidChecksumFormat);

  factory UnpQRParseError.checksumNotMatched() =>
      UnpQRParseError(UnpQRParseErrorType.checksumNotMatched);

  @override
  String toString() => 'UnpQRParseError(${type.name})';
}

/// UPN QR parse error type
enum UnpQRParseErrorType {
  /// Invalid format
  invalidFormat,

  /// Data is too long
  tooLong,

  /// Too few lines in the string data
  tooFewLines,

  /// Too many lines in the string data
  tooManyLines,

  /// There is no amount value in data
  missedAmount,

  /// There is no checksum value in data
  missedChecksum,

  /// Checksum format is invalid (not a correct number)
  invalidChecksumFormat,

  /// Checksum is not match.
  checksumNotMatched,
}


lib/src/upn_qr_parser.dart
import 'package:upn_qr/upn_qr.dart';

/// Parser utils for the UPN QR.
class UpnQRParser {
  // QR Character Set: ISO 8859-2. The use of Extended Channel Interpretation (ECI value 000004) is required.
  // At the end of each field, from sequence number 1 up to and including 20, add the punctuation LF (0x0A,
  // '\n'). If an individual field is blank, it contains only the punctuation mark.
  //
  // SN.  Field description         Max. length   QR content description
  // 1.   Leading style             5             Constant »UPNQR«.
  // 2.   Payer’s IBAN                            Rules for SI IBAN without spaces.
  // 3.   Deposit                   1             Uppercase "X" or blank.
  // 4.   Withdrawal                1             Uppercase "X" or blank
  // 5.   Payer’s reference         (4+22)        Model and reference together without spaces. 4 - SI or RF model rules.
  //                                              22 - Rules for SI or RF reference (without spaces
  // 6.   Payer’s name                            All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces
  // 7.   Payer’s street and No.                  All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces.
  // 8.   Payer’s city                            All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces.
  // 9.   Amount                                  Just numbers with leading zeros. Amount
  //                                              multiplied by 100 and without prefix or
  //                                              decimal places. The maximum amount
  //                                              that can be entered is
  //                                              999,999,999,999.99 (one billion - 1).
  //                                              If 0, the content is "00000000000"
  // 10.  Payment date                            Format 'DD.MM.YYYY'.
  // 11.  Urgent                    1             Uppercase "X" or blank. X X\n
  // 12.  Purpose code              4             Uppercase letters (A-Z). Purpose code rules.
  // 13.  Purpose of payment                      All characters of the ISO 8859-2 code table. No
  //                                              leading or following spaces.
  // 14.  Payment deadline                        Format 'DD.MM.YYYY'.
  // 15.  Recipient’s IBAN                        Rules for SEPA IBAN without spaces.
  // 16.  Recipient’s reference     (4+22)        Model and reference together without
  //                                              spaces. 4 - SI or RF model rules.
  //                                              22 - Rules for SI or RF reference (without spaces).
  // 17.  Recipient’s name                        All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces.
  // 18. Recipient’s street and No.               All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces.
  // 19.  Recipient’s city                        All characters of the ISO 8859-2 code
  //                                              table. No leading or following spaces.
  // 20.  Sum of the lengths of     3             Numbers with leading zeros.
  //      fields 1 to 19,                         Formula: Value = strlen(N1+'\n') + str-
  //      including punctuation                   len(N2+'\n') + ….. + strlen(N19+'\n').
  //      marks
  //      Reserve                                 Blank or spaces.
  //
  // (*) The reserve field may be blank or contain spaces and have no punctuation. The maximum length of the
  // reserve field is limited by the capacity of the QR record (V15, Binary, ECC_M, ECI ISO -8859-2), which is 411
  // characters. To complete the reserve field, the maximum length can be calculated using the following formula:
  // ReserveMaxLength = 411 - (strlen(N1+'\n') + strlen(N2+'\n') + ..... + strlen(N20+'\n')).
  //
  // The maximum length of the data, including punctuation without reserve, is 411 characters

  static const _maxLen = 411;

  /// Parse UPN QR code string data.
  static UpnData parseString(String value) {
    if (value.length > _maxLen) throw UnpQRParseError.tooLong();

    final lines = value.split('\n');
    if (lines.length > 21) throw UnpQRParseError.tooManyLines();
    if (lines.length < 20) throw UnpQRParseError.tooFewLines();

    var i = 0;
    if (lines[i++] != 'UPNQR') throw UnpQRParseError.invalidFormat();

    final payerIban = _parseStr(lines[i++]);
    final deposit = _parseBool(lines[i++]);
    final withdrawal = _parseBool(lines[i++]);
    final payerReference = _parseStr(lines[i++]);
    final payerName = _parseStr(lines[i++]);
    final payerAddress = _parseStr(lines[i++]);
    final payerCity = _parseStr(lines[i++]);
    final amount = _requireStr(lines[i++], UnpQRParseErrorType.missedAmount);
    final paymentDate = _parseStr(lines[i++]);
    final urgent = _parseBool(lines[i++]);
    final paymentPurposeCode = _parseStr(lines[i++]);
    final paymentPurpose = _parseStr(lines[i++]);
    final paymentDeadline = _parseStr(lines[i++]);
    final recipientIban = _parseStr(lines[i++]);
    final recipientReference = _parseStr(lines[i++]);
    final recipientName = _parseStr(lines[i++]);
    final recipientAddress = _parseStr(lines[i++]);
    final recipientCity = _parseStr(lines[i++]);

    final checkSum = int.tryParse(
        _requireStr(lines[i++], UnpQRParseErrorType.missedChecksum));
    if (checkSum == null) throw UnpQRParseError.invalidChecksumFormat();

    final calculatedChecksum =
        lines.take(19).fold(0, (v, l) => v + l.length + 1);

    if (calculatedChecksum != checkSum) {
      throw UnpQRParseError.checksumNotMatched();
    }

    return UpnData(
      payerIban: payerIban,
      deposit: deposit,
      withdrawal: withdrawal,
      payerReference: payerReference,
      payerName: payerName,
      payerAddress: payerAddress,
      payerCity: payerCity,
      amount: amount,
      paymentDate: paymentDate,
      urgent: urgent,
      paymentPurposeCode: paymentPurposeCode,
      paymentPurpose: paymentPurpose,
      paymentDeadline: paymentDeadline,
      recipientIban: recipientIban,
      recipientReference: recipientReference,
      recipientName: recipientName,
      recipientAddress: recipientAddress,
      recipientCity: recipientCity,
    );
  }

  static String? _parseStr(String value) => value.isEmpty ? null : value;

  static String _requireStr(String value, UnpQRParseErrorType errorType) =>
      value.isEmpty ? throw UnpQRParseError(errorType) : value;

  static bool _parseBool(String value) => value.toLowerCase() == 'x';

  UpnQRParser._();
}
